metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,423 @@
+      case termSelect: Term.Select
+          if termSelect.name.value == "flatMap" || termSelect.name.value == "map" =>
+        val qual = termSelect.qual
+        val qualString = document.substring(qual.pos.start, qual.pos.end)
+        termApply.args.head match {
+          case termFunction: Term.Function =>
+            extractForYieldOutOfFunction(
+              existingNameQuals,
+              existingConditions,
+              termFunction,
+              qualString,
+              document
+            )
+
+          case termBlock: Term.Block =>
+            termBlock.stats
+              .collectFirst { case termFunction: Term.Function =>
+                extractForYieldOutOfFunction(
+                  existingNameQuals,
+                  existingConditions,
+                  termFunction,
+                  qualString,
+                  document
+                )
+              }
+              .getOrElse(
+                (
+                  existingNameQuals,
+                  existingConditions,
+                  YieldExpression(
+                    document.substring(termApply.pos.start, termApply.pos.end)
+                  )
+                )
+              )
+
+          case _ =>
+            (
+              existingNameQuals,
+              existingConditions,
+              YieldExpression(
+                document.substring(termApply.pos.start, termApply.pos.end)
+              )
+            )
+        }
+      case _ =>
+        (
+          existingNameQuals,
+          existingConditions,
+          YieldExpression(
+            document.substring(termApply.pos.start, termApply.pos.end)
+          )
+        )
+    }
+  }
+
+  def extractForYieldOutOfPotentaialYieldTree(
+      existingNameQuals: List[ForYieldEnumeration],
+      existingConditions: List[ForYieldCondition],
+      yieldTree: Tree,
+      document: String
+  ): (List[ForYieldEnumeration], List[ForYieldCondition], YieldExpression) = {
+    yieldTree match {
+      case termApply: Term.Apply =>
+        extractForYield(
+          existingNameQuals,
+          existingConditions,
+          termApply,
+          document
+        )
+      case termIf: Term.If =>
+        extractForYieldOutOfIf(
+          existingNameQuals,
+          existingConditions,
+          termIf,
+          document
+        )
+      case termBlock: Term.Block if termBlock.stats.length == 1 =>
+        extractForYieldOutOfPotentaialYieldTree(
+          existingNameQuals,
+          existingConditions,
+          termBlock.stats.head,
+          document
+        )
+      case otherStat =>
+        (
+          existingNameQuals,
+          existingConditions,
+          YieldExpression(
+            document.substring(otherStat.pos.start, otherStat.pos.end)
+          )
+        )
+    }
+  }
+
+  private def extractForYieldOutOfIf(
+      existingNameQuals: List[ForYieldEnumeration],
+      existingConditions: List[ForYieldCondition],
+      termIf: Term.If,
```

Comment:
The `if ...` guard in for comprehensions is translated to `withFilter`, so this not exactly the same. We should only look for filter and not if.

Reply:
but `if` in all the scenarios I can think of also makes sense! Are you sure we want to remove it?

Reply:
Yes, it's a much more complex scenario that we need to understand the semantics instead of just syntax. You cannot know what can be returned as `else` and what should be returned in the main if branch. Option is one such possibility and it's rare that people do an if else branch instead of just filter.